packages/utilities/src/classNamesFunction.ts

```diff
@@ -1,32 +1,73 @@
-import { mergeStyleSets, IStyle, IStyleSet, IProcessedStyleSet } from '@uifabric/merge-styles';
+import { mergeStyleSets, IStyleSet, IProcessedStyleSet } from '@uifabric/merge-styles';
 import { IStyleFunctionOrObject } from '@uifabric/merge-styles';
 
+const MAX_CACHE_COUNT = 30;
+
+// Note that because of the caching nature within the classNames memoization,
+// I've disabled this rule to simply be able to work with any types.
+// tslint:disable:no-any
+
+const RetVal = '__retval__';
+
 /**
  * Creates a getClassNames function which calls getStyles given the props, and injects them
  * into mergeStyleSets.
+ *
+ * Note that the props you pass in on every render should be in the same order and
+ * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
+ * these will cause extra recalcs to occur.
  */
 export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(): (
   getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
   styleProps?: TStyleProps
 ) => IProcessedStyleSet<TStyleSet> {
-  // TODO: memoize.
+  let map: Map<any, any> = new Map<any, any>();
```

Comment:
should have comment explaining what this is a map of.. (what do the keys represent, what do values represent)

Reply:
Your scheme here is actually pretty cool but SERIOUSLY could use comments :) It also looks like the type can be better than `any, any`...

Reply:
will do for sure, thanks.